service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/kubernetesNamespace/DeleteKubernetesNamespaceStep.java

```diff
@@ -0,0 +1,105 @@
+package bio.terra.workspace.service.resource.controlled.cloud.azure.kubernetesNamespace;
+
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.common.utils.RetryUtils;
+import bio.terra.workspace.service.resource.controlled.cloud.azure.KubernetesClientProvider;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ControlledResourceKeys;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.openapi.apis.CoreV1Api;
+import io.kubernetes.client.openapi.models.V1NamespaceStatus;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+public class DeleteKubernetesNamespaceStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(DeleteKubernetesNamespaceStep.class);
+  private final UUID workspaceId;
+  private final KubernetesClientProvider kubernetesClientProvider;
+  private final ControlledAzureKubernetesNamespaceResource resource;
+
+  public DeleteKubernetesNamespaceStep(
+      UUID workspaceId,
+      KubernetesClientProvider kubernetesClientProvider,
+      ControlledAzureKubernetesNamespaceResource resource) {
+    this.workspaceId = workspaceId;
+    this.kubernetesClientProvider = kubernetesClientProvider;
+    this.resource = resource;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final AzureCloudContext azureCloudContext =
+        context
+            .getWorkingMap()
+            .get(ControlledResourceKeys.AZURE_CLOUD_CONTEXT, AzureCloudContext.class);
+
+    var coreApiClient =
+        kubernetesClientProvider.createCoreApiClient(azureCloudContext, workspaceId);
+
+    try {
+      logger.info("Deleting namespace {}", resource.getKubernetesNamespace());
+      coreApiClient.deleteNamespace(
+          resource.getKubernetesNamespace(), null, null, null, null, null, null);
+    } catch (ApiException e) {
+      return kubernetesClientProvider.stepResultFromException(e, HttpStatus.NOT_FOUND);
+    }
+
+    waitForDeletion(coreApiClient);
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  private void waitForDeletion(CoreV1Api coreApiClient)
+      throws InterruptedException, RetryException {
+    try {
+      RetryUtils.getWithRetry(
+          this::isNamespaceGone,
+          () -> getNamespaceStatus(coreApiClient),
+          Duration.ofMinutes(10),
+          Duration.ofSeconds(10),
+          0.0,
+          Duration.ofSeconds(10));
+    } catch (InterruptedException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new RuntimeException(
+          "timeout waiting for namespace %s to be deleted"
```

Comment:
How do we know for sure that the exception was caused by a timeout?

Reply:
the retry function is kind of crappy (it throws `Exception` so it is hard to disambiguate), I will see if I can make it better